public/app/features/datasources/settings/CustomHeadersSettings.tsx

```diff
@@ -0,0 +1,174 @@
+import React, { PureComponent } from 'react';
+import { Button, FormField, SecretFormField } from '@grafana/ui/src';
+
+enum HeaderChangeEvent {
+  add,
+  update,
+  reset,
+  remove,
+}
+
+export interface CustomHeader {
+  name: string;
+  value: string;
+  configured: boolean;
+}
+
+export type CustomHeaders = CustomHeader[];
+
+export interface Props {
+  datasource: {
+    jsonData: {
+      [key: string]: any;
+    };
+    secureJsonData: {
+      [key: string]: string;
+    };
+    secureJsonFields: {
+      [key: string]: boolean;
+    };
+  };
+  onChange: (headers: CustomHeaders) => void;
+}
+
+export interface State {
+  headers: CustomHeaders;
+}
+
+export class CustomHeadersSettings extends PureComponent<Props, State> {
+  state = {
+    headers: [],
+  };
+
+  constructor(props: Props) {
+    super(props);
+  }
+
+  componentDidMount(): void {
+    this.updateHeaders();
+  }
+
+  onHeaderChange(
+    type: HeaderChangeEvent,
+    e: React.ChangeEvent<HTMLInputElement> | React.MouseEvent | React.SyntheticEvent,
+    headerIndex?: number
+  ) {
+    e.preventDefault();
+    const eventTarget = e.target;
+    this.setState(
+      prevState => {
+        switch (type) {
+          case HeaderChangeEvent.add:
+            return { headers: [...prevState.headers, { name: 'X-Custom-Header', value: '', configured: false }] };
+          case HeaderChangeEvent.update:
+            const { name, value } = eventTarget as EventTarget & HTMLInputElement;
+            return { headers: prevState.headers.map(this.updateHeader(headerIndex, name, value)) };
+          case HeaderChangeEvent.reset:
+            return { headers: prevState.headers.map(this.resetHeader(headerIndex)) };
+          case HeaderChangeEvent.remove:
+            return {
+              headers: [...prevState.headers.slice(0, headerIndex), ...prevState.headers.slice(headerIndex + 1)],
+            };
+        }
+      },
+      () => this.props.onChange(this.state.headers)
```

Comment:
I think you should update the controller on input blur event instead of every change as until then you can think about the value as a transient state